test/common/tcp_proxy/upstream_test.cc

```diff
@@ -382,6 +382,32 @@ TYPED_TEST(HttpUpstreamRequestEncoderTest, RequestEncoderHeadersWithDownstreamIn
   auto ip_port = Network::Utility::getAddressWithPort(
       *Network::Test::getCanonicalLoopbackAddress(ip_versions[0]), 80);
   Network::ConnectionInfoSetterImpl connection_info(ip_port, ip_port);
+  EXPECT_CALL(this->downstream_stream_info_, downstreamAddressProvider)
+      .Times(2)
+      .WillRepeatedly(testing::ReturnRef(connection_info));
+  EXPECT_CALL(this->encoder_, encodeHeaders(HeaderMapEqualRef(expected_headers.get()), false));
+  this->upstream_->setRequestEncoder(this->encoder_, false);
+}
+
+TYPED_TEST(HttpUpstreamRequestEncoderTest,
+           RequestEncoderHostnameWithDownstreamInfoRequestedServerName) {
+  this->config_message_.set_hostname("%REQUESTED_SERVER_NAME%:443");
```

Comment:
Can we support arbitrary header additions? See https://github.com/envoyproxy/envoy/issues/21804. This is a major blocker for us after some recent changes landed. Given we can add arbitrary metadata for all headers _except_ `:authority` today it seems like this should be feasible?

Reply:
This change aims to allow setting command operators in the `tunneling_config.hostname` field to avoid setting empty or dummy hostname and then overriding header "Host" or ":authority" which are not allowed to override. But as @kyessenov requested, I'm refactoring this implementation to rely on substitution formatter which supports `%DYNAMIC_METADATA%`, so this change should provide what you need.